### PR TITLE
Introduces partitioner abstract base class while preserving old behav…

### PR DESCRIPTION
**Pull request by Bouncheck**
…iour and existing partitioner class. 
Introduces round robin partitioner as derived class.

### CODE REVIEW

File: src/kafka/utils/partitioner.hh [hunk 1]
```diff
@@ -32,10 +33,22 @@ class partitioner {
...
+class basic_partitioner : public partitioner {
+public:
+    virtual metadata_response_partition get_partition(const std::string &key, const kafka_array_t<metadata_response_partition> &partitions);
+};
+
+class rr_partitioner : public partitioner {
+public:
+    virtual metadata_response_partition get_partition(const std::string &key, const kafka_array_t<metadata_response_partition> &partitions);
+private:
+    std::atomic_uint32_t counter;
```
Comment (constantfold) [inline src/kafka/utils/partitioner.hh]: `atomic_uint32_t` is not necessary as producer runs on a single core/thread.

File: include/seastar/kafka/producer/kafka_producer.hh [hunk 2]
```diff
@@ -41,7 +41,7 @@ private:
 
     std::string _client_id;
     lw_shared_ptr<connection_manager> _connection_manager;
-    partitioner _partitioner;
+    partitioner* _partitioner = new basic_partitioner();
```
Comment (constantfold) [inline include/seastar/kafka/producer/kafka_producer.hh]: Move the initialization to `kafka_producer` constructor.

Currently it causes memory leak (`new` without `delete`) - maybe use `unique_ptr` instead?

File: include/seastar/kafka/producer/kafka_producer.hh [hunk 3]
```diff
@@ -41,7 +41,7 @@ private:
 
     std::string _client_id;
     lw_shared_ptr<connection_manager> _connection_manager;
-    partitioner* _partitioner = new basic_partitioner();
+    std::unique_ptr<partitioner> _partitioner;
```
Comment (StarostaGit) [inline include/seastar/kafka/producer/kafka_producer.hh:44]: Is there a reason for this to be allocated dynamically?

**Review by StarostaGit — COMMENTED** [review]
Left a question

**Review by StarostaGit — CHANGES_REQUESTED** [review]
This doesn't take care of the requests with a specific key given

File: include/seastar/kafka/producer/kafka_producer.hh (hunk 3 above)
Comment (Bouncheck) [reply include/seastar/kafka/producer/kafka_producer.hh:44]: It has to be this way if we want to use ABCs, because they cannot be instantiated.

**Review by StarostaGit — APPROVED** [review]
LGTM

**Pull request merged**